torchdata/stateful_dataloader/ibm_rescalable.py

```diff
@@ -0,0 +1,614 @@
+        logical_shard_states = [{k: v[i] for k, v in state_dict.items()} for i in range(self.n_logicals)]
+
+        # Load values
+        for i in range(self.n_logicals):
+            self.data[i].load_state_dict(logical_shard_states[i])
+
+
+#### -------------------------    CHECKPOINT FUNCTIONS    ------------------------- ####
+
+
+def __pop_dstate(state, device_mesh, placements):
+    """
+    Removes worker states from the StatefulDataLoader state dict, and assembles them
+    into a separate list of dicts for distributed checkpointing.
+    """
+    dstate = state["_snapshot"]["_worker_snapshots"]
+    dstate = [dstate[f"worker_{i}"].pop("dataset_state") for i in range(len(dstate))]
+    # Flip list[dict[tensor]] to dict[list[tensor]], and concat
+    dstate = {k: torch.cat([d[k] for d in dstate], 0) for k in dstate[0]}
+    # Construct dtensors from tensors
+    dstate = {
+        k: dtensor.DTensor.from_local(
+            v,
+            device_mesh,
+            placements,
+        )
+        for k, v in dstate.items()
+    }
+    return dstate
+
+
+def save_distributed_state_dict(
+    loader: StatefulDataLoader,
+    path: str,
+    device_mesh=None,
+):
+    """
+    Retrieves dataloader state dict, and separates worker states from loader state.
+    Loader state is not rescalable, and is saved using normal torch.save.
+    It is discarded when rescaling.
+    Rescalable worker states are compiled into a dtensor across ranks, and saved 
+    using pytorch distributed checkpointing.
+    """
+    rank = loader.dataset.rank
+    state = deepcopy(loader.state_dict())
+    dstate = __pop_dstate(state, device_mesh, [dtensor.placement_types.Shard(0)])
+    # Write distributed state dict
+    writer = checkpoint.FileSystemWriter(path)
+    checkpoint.save(
+        dstate,
+        writer,
+    )
+    # Write nondistributed state dict
+    torch.save(state, os.path.join(path, f"__nondist_cp_{rank}.pth"))
+
+
+def load_distributed_state_dict(
+    loader: StatefulDataLoader,
+    path: str,
+    device_mesh=None,
+):
+    """
+    Retrieves dataloader state dict, and separates worker states from loader state.
+    If not rescaling, load saved dataloader state.
+    Rescalable worker states are retrieved using pytorch distributed checkpointing.
+    States are distributed over workers, and ScalableShardDataset will handle
+    partitioning and re-assignment of available states into logical ranks.
+    """
+    base = loader.state_dict()
+    nworkers = base["_snapshot"]["_main_snapshot"]["_num_workers"]
+    rank = loader.dataset.rank
+    dstate = __pop_dstate(base, device_mesh, [dtensor.placement_types.Shard(0)])  # placements)
+    # Read nondistributed state dict
+    ckp_ws = 0 if not os.path.exists(path) else len([x for x in os.listdir(path) if "__nondist_cp_" in x])
+    # Check that number of loaders matches
+    if ckp_ws == loader.dataset.worldsize:
+        state = torch.load(os.path.join(path, f"__nondist_cp_{rank}.pth"))
```

Comment:
noob q: what are we missing out, if we just set the;

data_loader_state= base

without considering the rescaling property of the training run ?